utils/src/main/java/software/amazon/awssdk/utils/async/FlatteningSubscriber.java

```diff
@@ -68,34 +101,164 @@ public void cancel() {
+    }
+
+    /**
+     * Returns true if we need to call onNext downstream. If this is executed outside the handling-state-update condition, the
+     * result is subject to change.
+     */
+    private boolean onNextNeeded() {
+        return !allItems.isEmpty() && downstreamDemand.get() > 0;
+    }
+
+    /**
+     * If there are zero pending items in the queue and the upstream has called onComplete, then tell the downstream
+     * we're done.
+     */
+    private void handleOnCompleteState() {
+        if (onCompleteNeeded()) {
             subscriber.onComplete();
-        } else if (currentBatch.isEmpty() && demand.get() > 0) {
-            requestedNextBatch = true;
-            sourceSubscription.request(1);
+            terminalCallMadeDownstream = true;
         }
     }
 
-    @Override
-    public void onComplete() {
-        synchronized (lock) {
-            onCompleteCalled = true;
-            if (currentBatch.isEmpty()) {
-                subscriber.onComplete();
-            }
+    /**
+     * Returns true if we need to call onNext downstream. If this is executed outside the handling-state-update condition, the
+     * result is subject to change.
+     */
+    private boolean onCompleteNeeded() {
+        return allItems.isEmpty() && onCompleteCalledByUpstream && !terminalCallMadeDownstream;
+    }
+
+    /**
+     * If the upstream has called onError, then tell the downstream we're done, no matter what state the queue is in.
+     */
+    private void handleOnErrorState() {
+        if (onErrorNeeded()) {
+            subscriber.onError(onErrorFromUpstream);
+            terminalCallMadeDownstream = true;
+        }
+    }
+
+    /**
+     * Returns true if we need to call onError downstream. If this is executed outside the handling-state-update condition, the
+     * result is subject to change.
+     */
+    private boolean onErrorNeeded() {
+        return onErrorFromUpstream != null && !terminalCallMadeDownstream;
+    }
+
+    /**
+     * Ensure that we have at least 1 demand upstream, so that we can get more items.
+     */
+    private void ensureUpstreamDemandExists() {
+        if (this.upstreamDemand.get() < 0) {
+            log.error(() -> "Upstream delivered more data than requested. Resetting state to prevent a frozen stream.",
+                      new IllegalStateException());
+            upstreamDemand.set(1);
+            upstreamSubscription.request(1);
+        } else if (this.upstreamDemand.compareAndSet(0, 1)) {
+            upstreamSubscription.request(1);
+        }
+    }
+
+    private void die(Throwable e) {
+        log.error(() -> "Unexpected exception encountered that violates the reactive streams specification. Attempting to "
+                        + "terminate gracefully.", e);
+        try {
+            quietlyWithoutError(upstreamSubscription::cancel);
+            quietlyWithoutError(() -> subscriber.onError(e));
+        } finally {
+            terminalCallMadeDownstream = true;
```

Comment:
Would it make sense to set  `terminalCallMadeDownstream` first before cancel and onError? This way other methods relying on that flag can complete faster. I remember I've seen weird issues causing by the fact a flag is set after onComplete or onError which could take a long time to finish
Examples:
https://github.com/aws/aws-sdk-java-v2/commit/0b071010f3ff87936b3565693500b36a04c2b35b#diff-502a5d192387b30c3534d8f04fea3490175a24a7e38a07c5d5d2621e6d3e09bc
https://github.com/aws/aws-sdk-java-v2/commit/16f8627794715065c4de66248a53d5bb885925c5

Reply:
Sure. In this case we only ever mutate or read terminalCallMadeDownstream within the mutual exclusion flag so it shouldn't theoretically matter for ordering because it is all on one thread anyway, but if the downstream onComplete or onError throws an exception (which it should NEVER do), this would prevent duplicate onComplete and onError calls, which I guess is probably better?